educe/rst_dt/learning/features.py

```diff
@@ -2,390 +2,243 @@
-    """
-    Sparse features
-    """
-    _features =\
-        [MagicKey.basket_fn(ptb_pos_tags_in_first)]
-
-    def __init__(self):
-        desc = self.__doc__.strip()
-        super(PairSubgroup_Basket, self).__init__(desc, self._features)
-
-
-# export groups
-class SingleEduKeys(BaseSingleEduKeys):
-    """Single EDU features"""
-
-    def __init__(self):
-        groups = [
-            SingleEduSubgroup_Meta(),
-            SingleEduSubgroup_Text(),
-            SingleEduSubgroup_Ptb()
-        ]
-        super(SingleEduKeys, self).__init__(groups)
-
-
-class PairKeys(BasePairKeys):
-    """Features on a pair of EDUs"""
-
-    def __init__(self, sf_cache=None):
-        groups = [
-            PairSubGroup_Core(),
-            PairSubgroup_Gap(),
-            PairSubgroup_Tuple(sf_cache),
-            PairSubgroup_Basket()
-        ]
-        super(PairKeys, self).__init__(groups, sf_cache)
-
-    def init_single_features(self):
-        """Init features defined on single EDUs"""
-        return SingleEduKeys()
+PAIR_GAP = [
+    ('num_edus_between', Substance.CONTINUOUS),
+    ('same_paragraph', Substance.DISCRETE),
+    ('same_bad_sentence', Substance.DISCRETE),
+    ('same_ptb_sentence', Substance.DISCRETE)
+]
+
+
+def extract_pair_gap(edu_info1, edu_info2):
+    """Document tuple features"""
+    edu_num1 = edu_info1['edu'].num
+    edu_num2 = edu_info2['edu'].num
+
+    edu_num_diff = abs(edu_num2 - edu_num1) - 1
+    yield ('num_edus_between', edu_num_diff)
+
+    try:
+        para_id1 = edu_info1['para_idx']
+        para_id2 = edu_info2['para_idx']
+    except KeyError:
+        pass
+    else:
+        same_para = (para_id1 is not None and
+                     para_id2 is not None and
+                     para_id1 == para_id2)
+        yield ('same_paragraph', same_para)
+
+    # same sentence, raw segmentation
+    raw_sent_id1 = edu_info1['raw_sent_idx']
+    raw_sent_id2 = edu_info2['raw_sent_idx']
+    same_bad_sent = (raw_sent_id1 is not None and
+                     raw_sent_id2 is not None and
+                     raw_sent_id1 == raw_sent_id2)
+    yield ('same_bad_sentence', same_bad_sent)
+
+    # same sentence, ptb segmentation
+    # FIXME fix behaviour for sentences with no gold segmentation,
+    # as then edu2sent = edu2raw_sent
+    sent_id1 = edu_info1['sent_idx']
+    sent_id2 = edu_info2['sent_idx']
+    same_ptb_sent = (sent_id1 is not None and
+                     sent_id2 is not None and
+                     sent_id1 == sent_id2)
+    yield ('same_ptb_sentence', same_ptb_sent)
+
+
+PAIR_RAW_WORD = [
+    ('word_first_pairs', Substance.DISCRETE),
+    ('word_last_pairs', Substance.DISCRETE),
+    ('bigram_first_pairs', Substance.DISCRETE),
+    ('bigram_last_pairs', Substance.DISCRETE),
+]
+
+
+def extract_pair_raw_word(edu_info1, edu_info2):
+    """raw word features on EDU pairs"""
+    raw_words1 = edu_info1['raw_words']
+    raw_words2 = edu_info2['raw_words']
```

Comment:
I prefer to keep the flexibility of a dict for now as my long-term goal is to slice a document horizontally (each kind of information in an array, one cell per EDU) rather than vertically (an array of EDUs, one tuple per EDU).
This would enable to use numpy arrays with the numeric type appropriate for each kind of information (e.g. short uint for EDU id and sentence id, long uint for tokens if they are mapped from strings to uints...).